### PR TITLE
cephadm: reset cached SSH connection when host addr changes during re-add

### PR DESCRIPTION
**Pull request by ujjawal-ibm**
## Problem:
When a host already exists in the cluster, running `ceph orch host add <hostname> <new-ip>` succeeds even if `<new-ip>` belongs to a different machine. 
This happens because the hostname validation `check-host --expect-hostname` ends up using the existing SSH connection instead of connecting to the new address, so the new machine is never actually validated.

## Root cause:
The SSH connection is cached by hostname. When the same host is added again with a different IP address, `_check_valid_addr()` reuses the existing SSH connection to the old machine instead of creating a new connection to the new address. As a result, the hostname validation runs on the old machine and incorrectly succeeds.

## Fix:
check if the host already exists in inventory with a different address. If so, call self.ssh.reset_con(hostname) to drop the cached connection, forcing a fresh connection to the new address.
A try/except block around _check_valid_addr ensures that if validation fails (ex: hostname mismatch on the new IP), the bad connection created during validation is also cleaned up. Without this, the failed connection to the wrong address would persist in the cache and break subsequent operations for that host.

## Validation:
Re-add with wrong IP -> correctly rejected with hostname mismatch
Re-add with correct IP after failed attempt -> correctly accepted
Re-add with same IP + new labels -> correctly accepted

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -2215,7 +2215,20 @@ def _add_host(self, spec):
...
         HostSpec.validate(spec)
-        ip_addr = self._check_valid_addr(spec.hostname, spec.addr)
+        # Drop the cached SSH connection when re-adding a host with a new
+        # address. Otherwise check-host may reuse the old connection and
+        # validate the wrong machine.
+        addr_changed = spec.hostname in self.inventory and self.inventory.get_addr(spec.hostname) != spec.addr
+        if addr_changed:
+            self.ssh.reset_con(spec.hostname)
+        try:
+            ip_addr = self._check_valid_addr(spec.hostname, spec.addr)
```
Comment (ashjosh1git) [inline src/pybind/mgr/cephadm/module.py:2238]: 1. Pls check other callers for _check_valid_addr that may need a similar fix.
2. Additionally we can also check the scenario other way round where we try to add hosts having "same" actual hostnames. for example adding the following hosts:
  - Machine A at 192.168.1.10 with actual hostname webserver
  - Machine B at 192.168.1.20 with actual hostname webserver (same name)

Comment (ujjawal-ibm) [reply src/pybind/mgr/cephadm/module.py:2238]: 1. Checked all callers of _check_valid_addr, the only other one is update_host_addr, added similar change there as well.

2. this case is already handled by the existing check-host --expect-hostname validation. If the hostname provided to ceph orch host add doesnot match the machine's actual hostname then the validation will fail. and if two different machines actually have the same hostname, cephadm cannot distinguish between them bcoz it relies on the hostname reported by the machine itself. I think its more like a deployment level concern rather than something cephadm can guard againts/

**Comment by ujjawal-ibm** [comment]
@ashjosh1git I tested this scenario and without this fix, readding an existing host with a different I.P updates the host's address because the cached SSH connection is reused and the hostname validation runs against the old machine.
```
[root@ceph-node-0 ~]# ceph orch host ls
HOST         ADDR             LABELS      STATUS
ceph-node-0  192.168.100.100  _admin rgw
ceph-node-3  192.168.100.103              Offline
```
2 hosts in cluster
`[root@ceph-node-0 ~]# ceph orch host add ceph-node-0 192.168.100.102`
Added host ceph-node-0 with addr 192.168.100.102
```
[root@ceph-node-0 ~]# ceph orch host ls
HOST         ADDR             LABELS      STATUS
ceph-node-0  192.168.100.102  _admin rgw
ceph-node-3  192.168.100.103              Offline
```
**addr changed from .100 to .102**

With this fix, the cached connection is reset before validation, so it connects to the new I.P and fails with:
`ERROR: hostname "ceph-node-2" does not match expected hostname "ceph-node-0"`

```
[root@ceph-node-0 ~]# ceph orch host ls
HOST         ADDR             LABELS      STATUS
ceph-node-0  192.168.100.100  _admin rgw
ceph-node-3  192.168.100.103              Offline
```
```
[root@ceph-node-0 ~]# ceph orch host add ceph-node-0 192.168.100.102
Error EINVAL: check-host failed:
podman (/usr/bin/podman) version 4.2.0 is present
systemctl is present
lvcreate is present
Unit chronyd.service is enabled and running
ERROR: hostname "ceph-node-2" does not match expected hostname "ceph-node-0"
```
```
[root@ceph-node-0 ~]# ceph orch host ls
HOST         ADDR             LABELS      STATUS
ceph-node-0  192.168.100.100  _admin rgw
ceph-node-3  192.168.100.103              Offline

```
So our PR is handling this edge case.

**Comment by ujjawal-ibm** [comment]
jenkins test make check arm64

**Comment by ashjosh1git** [comment]
There could be a scenario where the cluster already has a host with same name as the host being added. The existing host may get its IP updated. Need not block this PR though. LGTM.

**Comment by ujjawal-ibm** [comment]
jenkins test make check arm64

**Review by ashjosh1git — APPROVED** [review]

**Comment by ShwetaBhosale1** [comment]
Teuthology run with this PR: https://pulpito.ceph.com/shwetabhosale-2026-08-11_13:19:31-orch:cephadm-wip-shweta-testing-2026-08-11-1539-distro-default-trial/

**Comment by ShwetaBhosale1** [comment]
ok to test
